lockbox-ios/Action/DataStoreAction.swift

```diff
@@ -0,0 +1,318 @@
+        self.openSubject
+                .take(1)
+                .flatMap { _ in
+                    self.checkState()
+                }
+                .flatMap { _ in
+                    self.webView.evaluateJavaScript("\(self.dataStoreName).list()")
+                }
+                .subscribe(onError:{ error in
+                    self.listSubject.onError(error)
+                })
+                .disposed(by: self.disposeBag)
+    }
+
+    private func checkState() -> Single<Bool> {
+        return _initialized().asObservable()
+                .flatMap { initialized -> Observable<Bool> in
+                    if !initialized {
+                        throw DataStoreError.NotInitialized
+                    }
+
+                    return self._locked().asObservable()
+                }
+                .map { locked -> Bool in
+                    if locked {
+                        throw DataStoreError.Locked
+                    }
+
+                    return locked
+                }
+                .asSingle()
+    }
+}
+
+extension DataStoreActionHandler: WKScriptMessageHandler, WKNavigationDelegate {
+    func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
+        self._open()
+                .take(1)
+                .subscribe(onError: { error in
+                    self.dispatcher.dispatch(action: ErrorAction(error: error))
+                    self.openSubject = ReplaySubject<Void>.create(bufferSize: 1)
+                 })
+                .disposed(by: self.disposeBag)
+    }
+
+    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
+        guard let function = JSCallbackFunction.init(rawValue: message.name) else {
+            self.dispatcher.dispatch(action: ErrorAction(error: DataStoreError.UnexpectedJavaScriptMethod))
+            return
+        }
+
+        switch function {
+        case .OpenComplete:
+            self.openSubject.onNext(())
+        case .InitializeComplete:
+            self.initializeSubject.onNext(())
+        case .UnlockComplete:
+            self.unlockSubject.onNext(())
+        case .LockComplete:
+            self.lockSubject.onNext(())
+        case .ListComplete:
+            guard let listBody = message.body as? [[Any]] else {
+                self.listSubject.onError(DataStoreError.UnexpectedType)
+                break
+            }
+
+            let itemList = listBody.flatMap { (anyList: [Any]) -> Item? in
+                guard let itemDictionary = anyList[1] as? [String: Any] else {
+                    return nil
+                }
+
+                var item:Item
+                do {
+                    item = try self.parser.itemFromDictionary(itemDictionary)
+                } catch {
+                    self.dispatcher.dispatch(action: ErrorAction(error: error))
+                    return nil
+                }
+
+                return item
+             }
```

Comment:
I was thinking nearer to:

```swift
let itemList = listBody.flatMap { (anyList: [Any]) -> Item? in
   guard let itemDictionary == anyList[1] as? [String: Any],
         let item = try? self.parser.itemFromDictionary(itemDictionary) else {
      // not sure which to do here.
      return nil
   }
   return item
}
```

I'm not sure what to do in the case of errors. You seem to be doing one thing in response to one malformed input, and another in another:

```swift
      // self.dispatcher.dispatch(action: ErrorAction(error: error))
      // self.listSubject.onError(DataStoreError.UnexpectedType)
```

I'm sure you'll want to handle the empty list case too.